pkg/checks/external/main.go

```diff
@@ -253,38 +253,57 @@ func (ext *Checker) getCheck() (*khcheckcrd.KuberhealthyCheck, error) {
 	return checkConfig, err
 }
 
-// cleanup cleans up any running checker pods by evicting them
+// cleanup cleans up any running, pending, or unknown checker pods by evicting them. Succeeded or Failed pods are left alone for records
+// if eviction fails, cleanup will attempt to forcefully kill the pod.
 func (ext *Checker) cleanup() {
 	ext.log("Evicting up any running pods with name", ext.podName())
 	podClient := ext.KubeClient.CoreV1().Pods(ext.Namespace)
 
 	// find all pods that are running still so we can evict them (not delete - for records)
 	checkLabelSelector := kuberhealthyCheckNameLabel + " = " + ext.CheckName
-	ext.log("eviction: looking for pods with the label", checkLabelSelector, "and status.phase=Running")
+	ext.log("eviction: looking for pods with the label", checkLabelSelector)
 	podList, err := podClient.List(metav1.ListOptions{
-		FieldSelector: "status.phase=Running",
 		LabelSelector: checkLabelSelector,
 	})
-	// if we cant list pods, just give up gracefully
+
+	// if we can't list pods, just give up gracefully
 	if err != nil {
 		ext.log("error when searching for checker pods to clean up", err)
 		return
 	}
 
-	// evict all checker pods found concurrently and exit
+	// evict all checker pods not status=Failed or status=Succeeded. Check for existense of pods afterwards and if eviction failed, forcefully attempt to kill the pods
 	wg := sync.WaitGroup{}
 	for _, p := range podList.Items {
-		wg.Add(1)
-		go func() {
-			ext.evictPod(p.GetName(), p.GetNamespace())
-			wg.Done()
-		}()
+		ext.log("finding pods that are not in status.phase=Failed or status.phase=Succeeded")
+		if p.Status.Phase != apiv1.PodFailed || p.Status.Phase != apiv1.PodSucceeded {
+			wg.Add(1)
+			go func(p apiv1.Pod) {
+				ext.log("evicting pod", p.GetName())
+				err := ext.evictPod(p.GetName(), p.GetNamespace())
+				if err != nil {
+					ext.log("pod eviction failed"+":", err)
+					ext.log("checking if pod", p.GetName, "still exists")
+					podExists, podErr := util.PodNameExists(ext.KubeClient, p.GetName(), p.GetNamespace())
+					if !podExists {
+						ext.log("Pod does not exist"+":", podErr)
+					} else {
+						ext.log("forcefull killing pod", p.GetName, "in 15 seconds")
+						err := util.PodKill(ext.KubeClient, p.GetName(), p.GetNamespace(), 15)
+						if err != nil {
+							ext.log("error killing pod", p.GetName()+":", err)
+						}
+					}
+				}
+				wg.Done()
+			}(p)
+		}
```

Comment:
Can we flatten this out or break it into a few supporting functions?

Reply:
I removed redundant logging since the evictPod, PodKill, and PodNameExists functions already do error logging and handling. The only other option I could think of was to combine PodNameExists and PodKill into the evictPod function but that seems it would defeat the purpose of breaking the functions out in the first place.

Reply:
Calls to util.PodNameExists and util.PodKill are now part of the evictPod func as discussed.